src/ser.rs

```diff
@@ -8,16 +8,129 @@
 use serde::{self, Deserialize, Deserializer, Serialize, Serializer};
 use std::string::ToString;
 use std_prelude::*;
+use std::fmt;
 use stfu8;
 
-use std::ffi::OsString;
+use super::{PathOps, PathMut};
+
+use std::ffi::{OsStr, OsString};
 #[cfg(unix)]
 use std::os::unix::ffi::{OsStrExt, OsStringExt};
 #[cfg(windows)]
 use std::os::windows::ffi::{OsStrExt, OsStringExt};
 
 use super::{PathAbs, PathDir, PathFile};
 
+#[derive(Clone, Eq, Hash, PartialEq, PartialOrd, Ord)]
+pub struct PathSer(Arc<PathBuf>);
+
+impl PathSer {
+    pub fn new<P: Into<Arc<PathBuf>>>(path: P) -> Self {
+        PathSer(path.into())
+    }
+}
+
+impl fmt::Debug for PathSer {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        self.0.fmt(f)
+    }
+}
+
+impl PathMut for PathSer {
+    fn append<P: AsRef<Path>>(&mut self, path: P) -> crate::Result<()> {
+        self.0.append(path)
+    }
+    fn pop_up(&mut self) -> crate::Result<()> {
+        self.0.pop_up()
+    }
+    fn truncate_to_root(&mut self) {
+        self.0.truncate_to_root()
+    }
+    fn set_file_name<S: AsRef<OsStr>>(&mut self, file_name: S) {
+        self.0.set_file_name(file_name)
+    }
+    fn set_extension<S: AsRef<OsStr>>(&mut self, extension: S) -> bool {
+        self.0.set_extension(extension)
+    }
+}
+
+impl PathOps for PathSer {
+    type Output = PathSer;
+
+    fn concat<P: AsRef<Path>>(&self, path: P) -> crate::Result<Self::Output> {
+        Ok(PathSer(self.0.concat(path)?))
+    }
+
+    fn with_file_name<S: AsRef<OsStr>>(&self, file_name: S) -> Self::Output {
+        PathSer(self.0.with_file_name(file_name))
+    }
+
+    fn with_extension<S: AsRef<OsStr>>(&self, extension: S) -> Self::Output {
+        PathSer(self.0.with_extension(extension))
+    }
+}
+
+impl AsRef<OsStr> for PathSer {
+    fn as_ref(&self) -> &std::ffi::OsStr {
+        self.0.as_ref().as_ref()
+    }
+}
+
+impl AsRef<Path> for PathSer {
+    fn as_ref(&self) -> &Path {
+        self.0.as_ref()
+    }
+}
+
+impl AsRef<PathBuf> for PathSer {
+    fn as_ref(&self) -> &PathBuf {
+        self.0.as_ref()
+    }
+}
+
+impl Borrow<Path> for PathSer {
+    fn borrow(&self) -> &Path {
+        self.as_ref()
+    }
+}
+
+impl Borrow<PathBuf> for PathSer {
+    fn borrow(&self) -> &PathBuf {
+        self.as_ref()
+    }
+}
+
+impl<'a> Borrow<Path> for &'a PathSer {
+    fn borrow(&self) -> &Path {
+        self.as_ref()
+    }
+}
+
+impl<'a> Borrow<PathBuf> for &'a PathSer {
+    fn borrow(&self) -> &PathBuf {
+        self.as_ref()
+    }
+}
+
+impl<P: AsRef<str>> From<P> for PathSer {
+    fn from(path: P) -> PathSer {
+        PathSer::new(PathBuf::from(path.as_ref()))
+    }
+}
+
+
+impl From<PathSer> for Arc<PathBuf> {
+    fn from(path: PathSer) -> Arc<PathBuf> {
+        path.0
+    }
+}
+
+impl From<PathAbs> for PathSer {
+    fn from(path: PathAbs) -> PathSer {
+        PathSer(path.0.clone())
```

Comment:
Do you need the `.clone()` here? Since we take ownership of the `PathAbs`, it's going to be de-allocated at the end of the method anyway, so we might as well re-use  the existing `Arc<PathBuf>` instead of creating a new one then immediately de-allocating the old one.

Reply:
I'm running clippy to catch things like this.

Also, note that it is only incrementing and then decrementing a counter, so not a lot of performance hit -- but ya, shouldn't be doing it :+1:

Reply:
wow, clippy didn't catch it. That might be worth a bug to clippy!

Reply:
Apparently this may be quite difficult for clippy to catch: https://github.com/rust-lang/rust-clippy/issues/17